### PR TITLE
WIP: Add DiscoveryInfo and custom domain name support

### PR DESCRIPTION
**Pull request by sttts**
- add domain name `Templates` configuration flag
- create non-canonical records according the domain patterns, interpolated as text/templates
- implement container IP proposal of https://github.com/mesosphere/mesos-dns/pull/197#issuecomment-128594330 (unfinished, see TODO below)

Reasons to implement our own template engine:
- text/template has no error handling for missing values of variables (until 5 months ago, but not in Go 1.3 or 1.4 yet)
- the template syntax is an interface to the user. If text/template causes performance problems (it seems to be pretty central in the record generation and could be very well be the bottleneck), we would not be able to optimize without the burden to fork the complete text/template library. With our own implementation we have full control.
- text/template would not allow easy checking for validity of a template
- text/template implements much more logic than we need, it uses reflection heavily. Overall it feels to be pretty heavy for our small domain name templates.

TODO:
- [x] re-add tests
- [x] drop records with missing context variable values
- [x] fallback from di-name -> name
- [x] document patterns
- [ ] add default values for variables in patterns (maybe follow-up PR)
- [ ] do benchmarks
- [ ] add (d) and (e) of container IP proposal at https://github.com/mesosphere/mesos-dns/pull/197#issuecomment-128594330 (maybe follow-up PR)
- [ ] add support for DiscoveryInfo *visibility* (maybe follow-up PR)
- [ ] use `domainFrag` and friends
- [ ] handle spec returning empty values
- [ ] use better hash function, e.g. `zbase32` (maybe follow-up PR)

### CODE REVIEW

File: records/generator.go [hunk 1]
```diff
@@ -21,12 +23,15 @@ import (
...
 
+var defaultDomainPatterns = []string{"{{.tname}}"}
+
 // Mesos-DNS state
 // Refactor when discovery id is available
 type RecordGenerator struct {
 	As     rrs
 	SRVs   rrs
 	Slaves map[string]string
+	DomainPatterns []string
```
Comment (jdef) [inline records/generator.go]: the current design passes Config items to each func as needed, why is this special? If RecordGenerator is going to house Config items, then perhaps it should declare Config as an anonymous field:
```go
type RecordGenerator struct {
  Config
  As rrs
  SRVs rrs
  //...
}
```

this allows:

1. simple Config assignment: `rg.Config = someOtherConfig`
1. simple access to embedded Config field's from RecordGenerator obj: `for _, pattern in rg.DomainPatterns {`
1. a more consistent initialization of default config values (`defaultDomainPatterns` probably belongs in Config with some of the other defaults)

File: records/generator.go [hunk 2]
```diff
@@ -232,24 +258,40 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
 				continue
 			}
 
-			tname := labels.AsDNS952(task.Name)
-			sid := slaveIdTail(task.SlaveId)
-			tag := hashString(task.Id)
-			tail := fname + "." + domain + "."
+			fields := map[string]string{
+				"sid": slaveIdTail(task.SlaveId),
```
Comment (jdef) [inline records/generator.go]: DRY: consider const values instead of repeating `"sid"`, et al. in several places

File: resolver/resolver_test.go [hunk 3]
```diff
@@ -242,7 +252,26 @@ func TestHandler(t *testing.T) {
...
+	res, err := fakeDNSwithPatterns(8053, &[]string{"test-{{.tname}}"})
+	if err != nil {
+		t.Error(err)
+	}
 
+	dns.HandleFunc("mesos.", res.HandleMesos)
+	go res.Serve("udp")
+	go res.Serve("tcp")
+
+	// wait for startup ? lame
```
Comment (jdef) [inline resolver/resolver_test.go]: there's a new way to wait for DNS server startup (`safeStartDNSResolver`) coming down the line in #144.
should probably do the same here once that PR has been merged.

File: records/generator.go (hunk 1 above)
Comment (sttts) [inline records/generator.go]: I have moved the pattern list out of the RecordGenerator again. Now it's a parameter to `InsertState` like all other settings coming from the config.

File: records/generator.go (hunk 2 above)
Comment (sttts) [inline records/generator.go]: This is not the final word yet. Next step is to identify the values we want in the patterns, now with the limited knowledge of the tasks, later from the `DiscoveryInfo` fields.

File: resolver/resolver_test.go (hunk 3 above)
Comment (sttts) [inline resolver/resolver_test.go]: Very good. Arbitrary sleeps to synchronize code hurt the eyes. What's the status of #146? Looks like your changes and mine will conflict, so we have to adapt this or #146, whichever is merged earlier.

Comment (jdef) [inline resolver/resolver_test.go]: I think I saw #146 merged today

File: records/generator.go [hunk 4]
```diff
@@ -21,6 +23,8 @@ import (
 // Will likely become map[string][]discoveryinfo
 type rrs map[string][]string
 
+var DefaultDomainPatterns = []string{"{{.TaskName}}"}
```
Comment (jdef) [inline records/generator.go]: I tend to dislike public, mutable, global defaults. Can we find another way?

Comment (sttts) [inline records/generator.go]: I also dislike it. Is there any better pattern than moving it into a function?

File: records/generator.go [hunk 5]
```diff
@@ -232,24 +258,41 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
...
 			}
 
-			tname := labels.AsDNS952(task.Name)
-			sid := slaveIdTail(task.SlaveId)
-			tag := hashString(task.Id)
-			tail := fname + "." + domain + "."
+			// context used to resolve DomainPattern variables for the current task
+			context := PatternContext{
+				SlaveId: slaveIdTail(task.SlaveId),
+				TaskId: hashString(task.Id),
```
Comment (jdef) [inline records/generator.go]: can we incorporate additional information into this hash? (and maybe modify the hash to actually use (at least) 32 bits). for example hashString(task.Id+task.SlaveId+task.frameworkId+task.executorId). and if we do that, do we really need the tail(slaveId) anymore?

File: records/generator.go [hunk 6]
```diff
@@ -232,24 +258,41 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
...
+
 			// SRV records
-			if task.Resources.Ports != "" {
+			if task.Resources.Ports != "" && trec != "" {
 				ports := yankPorts(task.Resources.Ports)
 				for _, port := range ports {
 					var srvhost string = trec + ":" + port
-					tcp := "_" + tname + "._tcp." + tail
-					udp := "_" + tname + "._udp." + tail
+					tcp := "_" + context.TaskName + "._tcp." + tail
```
Comment (jdef) [inline records/generator.go]: i'd actually like for us to shift our thinking re: using dynamic _service types. using standardized service names (even if they're our standard) can provide us some flexibility later on w/ respect to discovery. for example instead of the above:
```
_{taskName}._{proto}.{tail}
```
what about this:
```
_ep._{proto}.{taskName}.{tail}
```
and once we have service discovery, we can get create additional, named endpoint records:
```
_named._{proto}.{portName}.{taskName}.{tail}
--or--
_named._sub._ep._{proto}.{portName}.{taskName}.{tail}
```
(the `_ep` represents a task "endpoint")

or perhaps even this should be configurable via templates. thoughts?

File: records/generator.go (hunk 4 above)
Comment (jdef) [inline records/generator.go]: I'd be tempted to make it package-private and then write a public func that returns a clone. I'd still use the package-private reference within the package for ease of use and readability. And maybe a comment for the var that mentions that it's intended to be read-only.

Comment (sttts) [inline records/generator.go]: Made cloning method out of it.

File: records/generator.go (hunk 5 above)
Comment (sttts) [inline records/generator.go]: The `task.Id` is unique for the framework, the framework Id is unique for the cluster, compare https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L64

Right now we have the framework name in the domain, hence can leave out the framework Id from the hash and still get unique hashes.

I completely agree with 32 bits.

Comment (sttts) [inline records/generator.go]: Moreover, the reduction to 16 bits in `hashString` is not what is proposed for that hashing algorithm at http://www.isthe.com/chongo/tech/comp/fnv/index.html#xor-fold.

Comment (sttts) [inline records/generator.go]: Compare #157

File: records/generator.go (hunk 6 above)
Comment (sttts) [inline records/generator.go]: The question is whether we expect the users to do a 2-phase service discovery, i.e. querying the task names first with a template driven domain name schema and then the ports based on the result of the first query.

If not, we really should add the template driven, dynamic approach you sketched above.

I tend to prefer the second.

**Comment by sttts** [comment]
Rebased to build-fix branch. Examples are gone for now.

**Comment by sttts** [comment]
Based on #216. Only c1ffcb7 and 8765c8 matter here.

PS: what is github doing with the commit order. The commit view looks completely mixed up to me. Everything is fine in git-log, with the same sha1s.

File: docs/docs/configuration-parameters.md [hunk 7]
```diff
@@ -73,3 +74,5 @@ It is sufficient to specify just one of the `zk` or `masters` field. If both are
 `recurseon` controls if the DNS replies for names in the Mesos domain will indicate that recursion is available. The default value is `true`. 
 
 `enforceRFC952` will enforce an older, more strict set of rules for DNS labels. For details, see the [RFC-952](https://tools.ietf.org/html/rfc952). The default value is `false`.
+
+`templates` is a list of domain templates used to create records. The default value is `["{name}.{framework}"]`, compare the [naming documentation](naming.md).
```
Comment (tsenart) [inline docs/docs/configuration-parameters.md]: *domain name templates*

File: records/tmpl/templates.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+// Compile compiles a Template to a fast Compiled template.
+func (t Template) Compile(spec labels.Func) (*Compiled, error) {
+	tokens := []token{}
+
+	if string(t) == "" {
+		return nil, fmt.Errorf("invalid empty template")
+	}
+
+	// split template into tokens: strings, separators and {variables}
+	varRE, err := regexp.Compile(`{[\s\w-:]*}`)
```
Comment (jdef) [inline records/tmpl/templates.go]: my gut says that a regexp is overkill for the lexer. it's more code to hand-build a lexer but it will probably be more performant and you'll be able to provide better error messages to users when a template has problems.

Comment (sttts) [inline records/tmpl/templates.go]: This code is not performance criticial.

Comment (sttts) [inline records/tmpl/templates.go]: The execution of the templates is important code path performance-wise.

File: records/tmpl/templates.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+func (c *Compiled) Execute(ctx Context) (string, error) {
+	labels := make([]string, 0, len(c.tokens))
+	for _, t := range c.tokens {
+		label, err := t.interpolate(ctx)
+		if err != nil {
+			return "", err
+		}
+		labels = append(labels, label)
+	}
+	labels = append(labels)
```
Comment (jdef) [inline records/tmpl/templates.go]: why is this line needed?

File: docs/docs/naming.md [hunk 10]
```diff
@@ -56,6 +56,56 @@ Mesos-DNS generates A records for itself that list all the IP addresses that Mes
...
+```
+{name}.{framework}
+```
+
+This gives the A records described above, e.g. `nginx.marathon.mesos.`.
+
+Additional templates can be defined using a number of *template variables*:
+
+- `framework`: the framework name, e.g. `marathon`
+- `slave-id-short`: the number in the slave ID after the last dash `-`, e.g. `0` for a slave ID `20140803-125133-3041283216-5050-2410-0`
```
Comment (jdef) [inline docs/docs/naming.md]: what's the rationale behind the `-short` identifier? I don't think it's very robust.

File: docs/docs/naming.md [hunk 11]
```diff
@@ -56,6 +56,56 @@ Mesos-DNS generates A records for itself that list all the IP addresses that Mes
...
+- `slave-id`: the complete slave ID, e.g. `20140803-125133-3041283216-5050-2410-0`
+- `task-id`: the complate task ID, e.g. `chronos.49b91a9a-3dda-11e4-a088-c20493233aa5`
+- `task-id-hash`: a hash value of the task ID, e.g. 76830, sufficiently unique for a realistic number of tasks on a cluster
+- `name`: the name of a task, provided by the framework either in the *TaskInfo* or the *DiscoveryInfo*, e.g. `nginx`
+- `version`: the version of a task, provided by the framework in the *DiscoveryInfo*, e.g. `1.0`
+- `location`: the location of a task, provider by the framework in the *DiscoveryInfo*, e.g. `europe`
+- `environment`: the environment of a task, provider by the framework in the *DiscoveryInfo*, e.g. `prod`
+- `label:<key>`: the value for the given label key, provided by the framework in the *DiscoveryInfo*, e.g. label key `canary` with label value `lanzarote`
+
+The first 6 variables should exist for every task. The following variables only
```
Comment (jdef) [inline docs/docs/naming.md]: I think that all of this would be represented better in a table format vs. forcing me to count line by line through the above items.

File: records/tmpl/templates.go (hunk 9 above)
Comment (sttts) [inline records/tmpl/templates.go]: It's not. Good spot.

File: docs/docs/naming.md [hunk 12]
```diff
@@ -56,6 +56,56 @@ Mesos-DNS generates A records for itself that list all the IP addresses that Mes
...
+This gives the A records described above, e.g. `nginx.marathon.mesos.`.
+
+Additional templates can be defined using a number of *template variables*:
+
+- `framework`: the framework name, e.g. `marathon`
+- `slave-id-short`: the number in the slave ID after the last dash `-`, e.g. `0` for a slave ID `20140803-125133-3041283216-5050-2410-0`
+- `slave-id`: the complete slave ID, e.g. `20140803-125133-3041283216-5050-2410-0`
+- `task-id`: the complate task ID, e.g. `chronos.49b91a9a-3dda-11e4-a088-c20493233aa5`
+- `task-id-hash`: a hash value of the task ID, e.g. 76830, sufficiently unique for a realistic number of tasks on a cluster
+- `name`: the name of a task, provided by the framework either in the *TaskInfo* or the *DiscoveryInfo*, e.g. `nginx`
```
Comment (jdef) [inline docs/docs/naming.md]: assuming that the name in the DI trumps that of the task? if so, should clarify here

File: docs/docs/naming.md [hunk 13]
```diff
@@ -56,6 +56,56 @@ Mesos-DNS generates A records for itself that list all the IP addresses that Mes
...
+Mesos-DNS will interpolate the given list of *record templates* for each running
+task in the cluster. If a *referenced variable is not defined* or the value is
+the empty string, the record *will not be created*.
+
+Examples:
+- `{name}.{framework}` gives e.g. `nginx.marathon.mesos`.
+- `{name}.{location}.{framework}` gives e.g. `nginx.europe.marathon.mesos`, or no record at all for those tasks that don't define the location.
+- `{version}.{name}.{framework}` gives e.g. `1.0.nginx.marathon.mesos`, or no record at all for those tasks that don't define the version.
+- `{label:canary}.{name}` gives e.g. `lanzarote.nginx.mesos`.
+- `{label:unknown}.{name}` will give no record at all, if no label `unknown` is provided by the framework.
```
Comment (jdef) [inline docs/docs/naming.md]: what about labels with empty values?

File: records/generator.go [hunk 14]
```diff
@@ -344,42 +355,131 @@ func (rg *RecordGenerator) listenerRecord(listener string, ns string) {
...
+	compiledTmpl := compileNonCanonicalTemplates(templates, spec)
+	canonicalTmpl := compileEssentialTemplates("{name}-{task-id-hash}-{slave-id-short}.{framework}", spec)
+	tcpRFC2782Tmpl := compileEssentialTemplates("_{name}._tcp.{framework}", spec)
+	udpRFC2782Tmpl := compileEssentialTemplates("_{name}._udp.{framework}", spec)
+
+	// records to create
+	aRecordsTmpls := []recordTemplate{
+		{append(compiledTmpl, canonicalTmpl), domain, "task-ip"},
+		{append(compiledTmpl, canonicalTmpl), "slave." + domain, "slave-ip"},
+		{append(compiledTmpl, canonicalTmpl), "container." + domain, "container-ip"},
```
Comment (tsenart) [inline records/generator.go]: This doesn't look right. 3 appends of the same thing?

File: records/generator.go [hunk 15]
```diff
@@ -494,3 +588,29 @@ func getProto(pair string) (string, string, error) {
 	}
 	return h[0], h[1], nil
 }
+
+func specEachLabel(s string, spec labels.Func) string {
```
Comment (tsenart) [inline records/generator.go]: `labels.DomainFrag`?

File: records/generator.go [hunk 16]
```diff
@@ -494,3 +588,29 @@ func getProto(pair string) (string, string, error) {
...
+// return the slave number from a Mesos slave id
+func slaveIDTail(slaveID string) string {
+	fields := strings.Split(slaveID, "-")
+	return strings.ToLower(fields[len(fields)-1])
+}
+
+// BUG: The probability of hashing collisions is too high with only 17 bits.
+// NOTE: Using a numerical base as high as valid characters in DNS names would
+// reduce the resulting length without risking more collisions.
+func hashString(s string) string {
```
Comment (jdef) [inline records/generator.go]: you're probably better off not dropping bits and just base32-encoding the hash.

File: docs/docs/naming.md (hunk 13 above)
Comment (sttts) [inline docs/docs/naming.md]: See above: empty values => no record

File: records/generator.go (hunk 14 above)
Comment (sttts) [inline records/generator.go]: not elegant, but correct.

File: records/generator.go (hunk 16 above)
Comment (sttts) [inline records/generator.go]: It's pretty long then. Using a hash is fine, but the collision probability should be low for the cluster sizes we target. Compare generalized birthday paradox.

File: docs/docs/naming.md (hunk 10 above)
Comment (sttts) [inline docs/docs/naming.md]: It's still the old canonical task domain name. I agree it's not very clever.

File: records/generator.go (hunk 16 above)
Comment (jdef) [inline records/generator.go]: i'd actually prefer zbase32 (http://philzimmermann.com/docs/human-oriented-base-32-encoding.txt)

**Comment by sttts** [comment]
Closing. Follow-up PR #230 based on a local branch.
